tests/drivers/uart/uart_async_api/src/test_uart_async.c

```diff
@@ -584,7 +588,16 @@ ZTEST_USER(uart_async_read_abort, test_read_abort)
 
 	uart_tx(uart_dev, tx_buf, 95, 100 * USEC_PER_MSEC);
 
+#ifndef DUT_IS_UART_EMUL
 	k_timer_start(&read_abort_timer, K_USEC(300), K_NO_WAIT);
+#else
+	/* The uart emulation driver needs no real time to send and receive data. Any waiting here
+	 * will lead to the driver being finished with sending and receiving. So instead of using
+	 * the timer here, just disable receiving before the driver had any chance to start with
+	 * processing its tasks.
+	 */
+	uart_rx_disable(uart_dev);
+#endif
```

Comment:
This is also an indication that the emulated UART needs to be changed to behave like a real UART. Use a delayable work item to simulate receiving and transmitting taking time :)

Reply:
First of all, thank you for your review, i really appreciate this!

But I do not agree here. I feel that simulating the timing of each individual byte being sent or received is quite a bit of overkill and would add quite a bit of complexity to this. One could calculate when the events would arrive, but this feels even more complex. I think not having the correct timing here is acceptable, even when this means that the testcase needs a small change as a compromise.

When i think about, what i want to achieve as a developer when using this fixture, i usually want to test if my code did use the API correctly and does work with the events correctly. So the events need to come in in the correct order, have the correct payloads etc. Timing is not that big of an issue, i think. E.g. if one needs the code under test to receive a message at an exact point in time, one could simply provide the data with `uart_emul_put_rx_data()` at the correct moment in time.

Also the already existing code does not do that, the implementations of the polling and the interrupt based APIs do not respect accurate timing either.

Reply:
That sounds more like a mock than an emulated UART, something better implemented using the FFF framework I think. Emulated device drivers are typically used to test the test suites themselves, and to provide a "fake" device that can be used to develop applications, like the emulated RTC and GNSS device drivers.

If the emulated UART is meant to be used as a mock, which it seems like it is, it should not be tested as part of the test suite for real UARTs, as it is not one (nor emulating one)

I would simply remove the emulated UART from the uart async api test suite, WDYT?

Reply:
Thats the other extreme. I would like to have it behave like a real UART  at the API level as much as possible and the already existing test cases come in handy to test this, as those test zephyrs expectation on how the API should behave. I also got to jump through some extra hoops to make it to comply with those test cases as much as possible. So i think its a bit more than a simple mock. I just don't wanna over engineer it.

Reply:
Not sure how to resolve this. I would be quite unhappy with removing it from this tests and i also do not want add simulation of the timings here, as this would add quite a bit of complexity and needs a bit of work to do for unclear benefit.

Reply:
The benefit is avoiding violating the UART API and test suite by coupling them to a specific device driver which is not behaving like a real UART, which is what the API and test suite are designed to abstract and validate.

Yes it is more complexity to make it behave like a real UART, but that in itself is not a sufficient argument against adding said complexity if the goal is to make it emulate a real UART. If the goal is not to emulate a real UART, don't add the additional complexity, and don't treat it like a real UART by for example adding it to the UART test suite :)

I'm not saying which is the better option, but those are the options AFAIK

Reply:
I do think it is better not to modify the tests to suit a particular driver.

Reply:
I dropped the commit enabling these tests now.

I'm not very happy about this, as i saw some value in validating the API as far as possible. But i do not have the time to write a complete simulation of a uart peripheral. I wrote this driver as part of a project to have a nice fixture for testing code using the uart async API and thought it might be useful for others, too. I think testing that the API does help to increase confidence in those tests.